### PR TITLE
[SYCL][UR] Update unified-runtime

### PR DESCRIPTION
**Pull request by jandres742**
Update to

https://github.com/oneapi-src/unified-runtime/commit/61cb864b3a73918124bff11f9ad58dafb8c7769a

Signed-off-by: Jaime Arteaga <jaime.a.arteaga.molina@intel.com>

### CODE REVIEW

**Comment by jandres742** [comment]
SYCL / Linux / Build + LIT (pull_request_target)  https://github.com/intel/llvm/actions/runs/4154130558/jobs/7187700934 failing with:

```cpp
_deps/unified-runtime-src/source/loader/ur_loader.cpp:36:2: error: extra ';' [-Werror=pedantic]
   36 | };
      |  ^
_deps/unified-runtime-src/source/loader/ur_loader.cpp:43:2: error: extra ';' [-Werror=pedantic]
   43 | };
```

Posted PR in UR https://github.com/oneapi-src/unified-runtime/pull/254

File: sycl/plugins/unified_runtime/ur/ur.hpp [hunk 1]
```diff
@@ -356,13 +355,13 @@ class UrReturnHelper {
...
 
 // Utility function for setting a message and warning
 [[maybe_unused]] void setErrorMessage(const char *message,
-                                      zer_result_t error_code);
+                                      ur_result_t error_code);
 
 // Returns plugin specific error and warning messages
 // TODO: promote to Unified Runtime API
-zer_result_t zerPluginGetLastError(char **message);
+ur_result_t zerPluginGetLastError(char **message);
```
Comment (igchor) [inline sycl/plugins/unified_runtime/ur/ur.hpp:367]: I think you don't need this function anymore. You can just implement `urGetLastResult`.
Feel free to ignore this comment if this is something you plan to do in next PRs.

Comment (jandres742) [reply sycl/plugins/unified_runtime/ur/ur.hpp:367]: Thanks @igchor . good point. will remove->implement in follow up patch.

**Comment by jandres742** [comment]
I have identified some issues in tests. Fixing them in my machine and will repost.

**Comment by jandres742** [comment]
@smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415

All I see is are errors like:

rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied

**Comment by smaslov-intel** [comment]
> @smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415
> 
> All I see is are errors like:
> 
> rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied

I don't see either. Did you try to restart it?
@bader : any idea?

**Comment by bader** [comment]
> > @smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415
> > All I see is are errors like:
> > rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied
> 
> I don't see either. Did you try to restart it? @bader : any idea?

No.

**Comment by bader** [comment]
> @smaslov-intel : do you see why this is failing? https://github.com/intel/llvm/actions/runs/4191948209/jobs/7266948415
> 
> All I see is are errors like:
> 
> rm: cannot remove '/__w/llvm/llvm/bolt/test/shared-object.test': Permission denied

@intel/dpcpp-devops-reviewers, could you investigate, please? To reduce the impact on CI jobs we probably should disable broken systems in GitHub Actions.

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 2]
```diff
@@ -1,31 +1,63 @@
...
 endif()
 
+add_library (UnifiedRuntimeLoader INTERFACE)
+get_filename_component(UNIFIED_RUNTIME_LIBRARY_SRC "${UNIFIED_RUNTIME_LOADER_SOURCE_DIR}" DIRECTORY)
+get_filename_component(UNIFIED_RUNTIME_LIB_NAME "${UNIFIED_RUNTIME_LIBRARY}" NAME)
+
+message(STATUS "UNIFIED_RUNTIME_LIB_NAME ${UNIFIED_RUNTIME_LIB_NAME}")
+
 add_library (UnifiedRuntime-Headers INTERFACE)
+add_library (UnifiedRuntimeLoader-Headers INTERFACE)
```
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/CMakeLists.txt:42]: do we really need two different headers interfaces here? why?

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 3]
```diff
@@ -1,31 +1,63 @@
...
+  set(UNIFIED_RUNTIME_LIBRARY loader)
   set(UNIFIED_RUNTIME_SOURCE_DIR
     ${unified-runtime_SOURCE_DIR} CACHE PATH "Path to Unified Runtime Headers")
+  set(UNIFIED_RUNTIME_LOADER_SOURCE_DIR
+    ${unified-runtime_SOURCE_DIR}/source/loader CACHE PATH "Path to Unified Runtime Loader Headers")
   set(UNIFIED_RUNTIME_INCLUDE_DIR "${UNIFIED_RUNTIME_SOURCE_DIR}/include")
 endif()
 
+add_library (UnifiedRuntimeLoader INTERFACE)
+get_filename_component(UNIFIED_RUNTIME_LIBRARY_SRC "${UNIFIED_RUNTIME_LOADER_SOURCE_DIR}" DIRECTORY)
```
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/CMakeLists.txt:36]: `UNIFIED_RUNTIME_LIBRARY_SRC` seems unused

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 4]
```diff
@@ -52,9 +84,30 @@ add_sycl_plugin(unified_runtime
...
+  LIBRARIES
     Threads::Threads
     UnifiedRuntime-Headers
+    UnifiedRuntimeLoader-Headers
+    UnifiedRuntimeLoader
     LevelZeroLoader-Headers
     LevelZeroLoader
 )
+
+if (TARGET loader)
```
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/CMakeLists.txt:96]: I assume this `loader` target is from UR repo, right? Is it possible to name it so it is more clear?

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 5]
```diff
@@ -1,31 +1,63 @@
...
+
 add_library (UnifiedRuntime-Headers INTERFACE)
+add_library (UnifiedRuntimeLoader-Headers INTERFACE)
+
 target_include_directories(UnifiedRuntime-Headers
   INTERFACE
     "${UNIFIED_RUNTIME_INCLUDE_DIR}"
 )
 
+target_include_directories(UnifiedRuntimeLoader-Headers
```
Comment (igchor) [inline sycl/plugins/unified_runtime/CMakeLists.txt:49]: I'm not sure if those `target_include_directories` and `target_link_libraries` and those extra variables (UNIFIED_RUNTIME_LOADER_SOURCE_DIR, UNIFIED_RUNTIME_INCLUDE_DIR, etc.) are actually needed.

UR provides a target named unified-runtime::loader which already defines include dirs, so I think it should be enough to just link with `unified-runtime::loader` (put it under `LIBRARIES`  in `add_sycl_plugin` below)

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 6]
```diff
@@ -52,9 +84,30 @@ add_sycl_plugin(unified_runtime
     # These below belong to Unified Runtime PI Plugin only
     "pi_unified_runtime.hpp"
     "pi_unified_runtime.cpp"
-    LIBRARIES
+  LIBRARIES
     Threads::Threads
     UnifiedRuntime-Headers
+    UnifiedRuntimeLoader-Headers
+    UnifiedRuntimeLoader
     LevelZeroLoader-Headers
```
Comment (igchor) [inline sycl/plugins/unified_runtime/CMakeLists.txt:92]: I think we should link only with the UnifiedRuntimeLoader here. And the L0 adapter should be build as a standalone SHARED library that links with LevelZero loader. 

Something like this:
```
add_sycl_plugin(unified_runtime
  SOURCES
    "pi_unified_runtime.hpp"
    "pi_unified_runtime.cpp"
    LIBRARIES
    Threads::Threads
    UnifiedRuntimeLoader
    UnifiedRuntime-Headers
)
```

and:
```
add_library(ur_level_zero
    SHARED
        ${CMAKE_CURRENT_SOURCE_DIR}/ur_level_zero.cpp
        ${CMAKE_CURRENT_SOURCE_DIR}/ur_level_zero.hpp
        "../../../ur_bindings.hpp"
        "../../../pi2ur.hpp"
        "../../../pi2ur.cpp"
        "../../ur.hpp"
        "../../ur.cpp"
)

target_include_directories(ur_level_zero
  PRIVATE "${LEVEL_ZERO_INCLUDE_DIR}"
)

target_link_libraries(ur_level_zero
  PRIVATE
    LevelZeroLoader-Headers
    LevelZeroLoader
)
```

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 3 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/CMakeLists.txt:36]: @smaslov-intel : I added several changes here to start using the loader. ur_loader.so is now being created but not used yet. These extra instructions are me trying to use the ur_loader, but decided to focus on code and then continue with the loader later.

I can remove all these if you'd like.

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 2 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/CMakeLists.txt:42]: please see above.

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 5 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/CMakeLists.txt:49]: same as above.

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 4 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/CMakeLists.txt:96]: @smaslov-intel loader is the name being used at this moment in the UR. I could submit a PR to the UR to change the name.

**Comment by jandres742** [comment]
the changes in the CMakeLlists for the loader are not  needed for the update and rest of changes. I could separate this into two patches, one for the changes in CMakeLists for the loader, and other for the changes in code.

@smaslov-intel , @igchor : what do you think?

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 4 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: @smaslov-intel : used alias now, please see.

**Comment by jandres742** [comment]
> the changes in the CMakeLlists for the loader are not needed for the update and rest of changes. I could separate this into two patches, one for the changes in CMakeLists for the loader, and other for the changes in code.
> 
> @smaslov-intel , @igchor : what do you think?

@smaslov-intel , @igchor : I have separated the changes in CMake for the loader in https://github.com/intel/llvm/pull/8423. I will incorporate your comments there, and this PR will be mostly about updating UR version.

**Review by smaslov-intel — APPROVED** [review]
Thanks! LGTM

> [jandres742](https://github.com/jandres742) [force-pushed](https://github.com/intel/llvm/compare/d580073eb4457f93055116a2e85a47da92cf09e4..45bd81e5248a5619030837a85253720221dda4d0) the updateur branch 2 times, most recently from [d580073 ](https://github.com/intel/llvm/commit/d580073eb4457f93055116a2e85a47da92cf09e4)to [45bd81e ](https://github.com/intel/llvm/commit/45bd81e5248a5619030837a85253720221dda4d0)[14 hours ago](https://github.com/intel/llvm/pull/8315#event-8579104475)

@jandres742: please try to avoid **force-push** in future as it makes incremental review impossible

**Comment by jandres742** [comment]
A first stable tag has been added to the UR. 

I need in any case to rebase due to conflicts. Should I:

1. rebase as is, and then submit a new patch updating to the stable tag, or?
2. update to stable tag and use this same PR?

I'm thinking to update to that commit and use that for now while we work on updating L0 adapter.

Please let me know what everyone thinks.

@smaslov-intel , @bader  @igchor

**Comment by bader** [comment]
> I need in any case to rebase due to conflicts.

Please, use `git merge <remote>/sycl` command to resolve merge conflicts. 

To update the tag/source/<make any other change> just push another commit to your branch. 
Don't worry about the history in your branch. All commits in this PR will be squashed to keep the linear history in the `sycl` branch. 
https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#pull-request

**Comment by jandres742** [comment]
Rebased and updated to UR tag https://github.com/oneapi-src/unified-runtime/tree/weekly-2023-02-24

Idea is to use this stable tag for development for now.

**Review by smaslov-intel — APPROVED** [review]

**Review by steffenlarsen — APPROVED** [review]
Just a small nitpick, otherwise changes look good to me.

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 7]
```diff
@@ -58,3 +71,7 @@ add_sycl_plugin(unified_runtime
...
     LevelZeroLoader
 )
+
+if (TARGET ${PROJECT_NAME}::ur_loader)
+  set_target_properties(hello_world PROPERTIES EXCLUDE_FROM_ALL 1 EXCLUDE_FROM_DEFAULT_BUILD 1)
+endif()
```
[the remark this answers is not in the document]
Comment (jandres742) [reply sycl/plugins/unified_runtime/CMakeLists.txt:77]: thanks @steffenlarsen . changing.

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 6 above)
Comment (jandres742) [reply sycl/plugins/unified_runtime/CMakeLists.txt:71]: ur_loader work to continue in https://github.com/intel/llvm/pull/8423

**Review by igchor — APPROVED** [review]
LGTM

**Pull request closed without merging**